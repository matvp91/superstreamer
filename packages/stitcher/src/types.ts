import type { DateTime } from "luxon";

export type InterstitialAssetType = "ad" | "bumper";

export interface InterstitialAsset {
  url: string;
  type?: InterstitialAssetType;
}

export interface Interstitial {
  dateTime: DateTime;
  vastUrl?: string;
  vastData?: string;
  asset?: InterstitialAsset;
  assetListUrl?: string;
}
