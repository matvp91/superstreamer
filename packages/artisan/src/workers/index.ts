import { runWorkers } from "bolt";
import { ffmpegCallback } from "./ffmpeg";
import { ffprobeCallback } from "./ffprobe";
import { imageCallback } from "./image";
import { packageCallback } from "./package";
import { pipelineCallback } from "./pipeline";
import { transcodeCallback } from "./transcode";

runWorkers([
  {
    name: "transcode",
    callback: transcodeCallback,
  },
  {
    name: "package",
    callback: packageCallback,
  },
  {
    name: "ffmpeg",
    callback: ffmpegCallback,
  },
  {
    name: "ffprobe",
    callback: ffprobeCallback,
  },
  {
    name: "pipeline",
    callback: pipelineCallback,
  },
  {
    name: "image",
    callback: imageCallback,
  },
]);
