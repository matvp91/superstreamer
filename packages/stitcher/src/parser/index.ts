export { parseMasterPlaylist, parseMediaPlaylist } from "./parse";
export { stringifyMasterPlaylist, stringifyMediaPlaylist } from "./stringify";
export { getRenditions } from "./helpers";

export * from "./types";
