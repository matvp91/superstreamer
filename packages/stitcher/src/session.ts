import { randomUUID } from "crypto";
import { DateTime } from "luxon";
import { kv } from "./adapters/kv";
import { JSON } from "./lib/json";
import { resolveUri } from "./lib/url";
import { fetchVmap } from "./vmap";
import type { Interstitial, InterstitialType } from "./interstitials";
import type { VmapParams, VmapResponse } from "./vmap";

export interface Session {
  id: string;
  url: string;
  expiry: number;

  startTime?: DateTime;

  // User defined options
  vmap?: VmapParams;
  vmapResponse?: VmapResponse;
  interstitials?: Interstitial[];
}

export async function createSession(params: {
  uri: string;
  vmap?: {
    url: string;
  };
  interstitials?: {
    position: number;
    uri: string;
    duration?: number;
    type?: InterstitialType;
  }[];
  expiry?: number;
}) {
  const id = randomUUID();

  const session: Session = {
    id,
    url: resolveUri(params.uri),
    vmap: params.vmap,
    // A session is valid for 3 hours by default.
    expiry: params.expiry ?? 60 * 60 * 3,
  };

  if (params.interstitials) {
    session.interstitials = params.interstitials.map((interstitial) => {
      return {
        position: interstitial.position,
        url: resolveUri(interstitial.uri),
        duration: interstitial.duration,
        type: interstitial.type,
      };
    });
  }

  // We'll initially store the session for 10 minutes, if it's not been consumed
  // within the timeframe, it's gone.
  const value = JSON.stringify(session);
  await kv.set(`session:${id}`, value, 60 * 10);

  return session;
}

export async function getSession(id: string) {
  const data = await kv.get(`session:${id}`);
  if (!data) {
    throw new Error(`No session found for id ${id}`);
  }
  return JSON.parse<Session>(data);
}

export async function processSessionOnMasterReq(session: Session) {
  // Check if we have a startTime, if so, the master playlist has been requested
  // before and we no longer need it.
  if (session.startTime) {
    return;
  }

  session.startTime = DateTime.now();

  if (session.vmap) {
    session.vmapResponse = await fetchVmap(session.vmap);
    delete session.vmap;
  }

  const value = JSON.stringify(session);
  await kv.set(`session:${session.id}`, value, session.expiry);
}
