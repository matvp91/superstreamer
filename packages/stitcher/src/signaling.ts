import type { InterstitialAsset } from "./types";

interface SignalingEvent {
  type: "impression" | "quartile" | "clickthrough";
  start?: number;
  urls: string[];
}

const QUARTILE_EVENTS: Record<string, number> = {
  start: 0,
  firstQuartile: 0.25,
  midpoint: 0.5,
  thirdQuartile: 0.75,
  complete: 1,
};

export function getSignalingForAsset(
  assets: InterstitialAsset[],
  asset: InterstitialAsset,
) {
  const { duration, tracking } = asset;
  if (!tracking) {
    return null;
  }

  const assetIndex = assets.indexOf(asset);
  const startTime = assets.splice(0, assetIndex).reduce((acc, asset) => {
    acc += asset.duration;
    return acc;
  }, 0);

  const signalingEvents: SignalingEvent[] = [];

  signalingEvents.push({
    type: "impression",
    start: 0,
    urls: tracking.impression,
  });

  signalingEvents.push({
    type: "clickthrough",
    urls: tracking.clickThrough,
  });

  // Map each tracking URL to their corresponding quartile.
  Object.entries(tracking).forEach(([name, urls]) => {
    const offset = QUARTILE_EVENTS[name];
    if (offset !== undefined) {
      signalingEvents.push({
        type: "quartile",
        start: duration * offset,
        urls,
      });
    }
  });

  return {
    version: 2,
    type: "slot",
    payload: [
      {
        type: "linear",
        start: startTime,
        duration: asset.duration,
        tracking: signalingEvents,
      },
    ],
  };
}
