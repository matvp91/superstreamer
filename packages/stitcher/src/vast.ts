import { DOMParser } from "@xmldom/xmldom";
import { AudioCodec, VideoCodec } from "bolt";
import * as uuid from "uuid";
import { VASTClient } from "vast-client";
import { api } from "./lib/api-client";
import type { VmapAdBreak } from "./vmap";
import type { VastAd, VastCreativeLinear, VastResponse } from "vast-client";

const NAMESPACE_UUID_AD = "5b212a7e-d6a2-43bf-bd30-13b1ca1f9b13";

export interface AdMedia {
  assetId: string;
  fileUrl: string;
  duration: number;
}

export async function getAdMediasFromAdBreak(adBreak: VmapAdBreak) {
  const adMedias = await getAdMedias(adBreak);
  const result: AdMedia[] = [];

  for (const adMedia of adMedias) {
    const asset = await fetchAsset(adMedia.assetId);
    if (!asset) {
      await scheduleForPackage(adMedia);
    } else {
      // If we have an asset registered for the ad media,
      // add it to the result.
      result.push(adMedia);
    }
  }

  return result;
}

async function getAdMedias(adBreak: VmapAdBreak): Promise<AdMedia[]> {
  const vastClient = new VASTClient();
  const parser = new DOMParser();

  let vastResponse: VastResponse | undefined;

  if (adBreak.vastUrl) {
    vastResponse = await vastClient.get(adBreak.vastUrl);
  } else if (adBreak.vastData) {
    const xml = parser.parseFromString(adBreak.vastData, "text/xml");
    vastResponse = await vastClient.parseVAST(xml);
  }

  if (!vastResponse) {
    return [];
  }

  return await formatVastResponse(vastResponse);
}

async function scheduleForPackage(adMedia: AdMedia) {
  await api.pipeline.post({
    assetId: adMedia.assetId,
    group: "ad",
    inputs: [
      {
        path: adMedia.fileUrl,
        type: "video",
      },
      {
        path: adMedia.fileUrl,
        type: "audio",
        language: "eng",
      },
    ],
    streams: [
      {
        type: "video",
        codec: VideoCodec.h264,
        height: 720,
      },
      {
        type: "video",
        codec: VideoCodec.h264,
        height: 480,
      },
      {
        type: "audio",
        codec: AudioCodec.aac,
        language: "eng",
      },
    ],
  });
}

async function fetchAsset(id: string) {
  const { data, status } = await api.assets({ id }).get();
  if (status === 404) {
    return null;
  }
  if (status === 200) {
    return data;
  }
  throw new Error(`Failed to fetch asset, got status ${status}`);
}

async function formatVastResponse(response: VastResponse) {
  return response.ads.reduce<AdMedia[]>((acc, ad) => {
    const creative = getCreative(ad);
    if (!creative) {
      return acc;
    }

    const mediaFile = getMediaFile(creative);
    if (!mediaFile?.fileURL) {
      return acc;
    }

    const adId = getAdId(creative);

    acc.push({
      assetId: adId,
      fileUrl: mediaFile.fileURL,
      duration: creative.duration,
    });

    return acc;
  }, []);
}

function getMediaFile(creative: VastCreativeLinear) {
  const mediaFiles = creative.mediaFiles
    .filter((mediaFile) => mediaFile.mimeType === "video/mp4")
    .sort((a, b) => b.height - a.height);
  return mediaFiles[0] ?? null;
}

function getCreative(ad: VastAd) {
  for (const creative of ad.creatives) {
    if (creative.type === "linear") {
      return creative as VastCreativeLinear;
    }
  }
  return null;
}

function getAdId(creative: VastCreativeLinear) {
  if (creative.adId && creative.id) {
    // Do not change this, or we'll have a mismatch between the already encoded ad's and the other.
    // See https://iabtechlab.com/guidance-for-uniquely-identifying-creative-asset-ids-in-vast-2/
    const adId = [creative.adId, creative.id].join(".");
    return uuid.v5(adId, NAMESPACE_UUID_AD);
  }

  throw new Error("Failed to generate adId");
}
