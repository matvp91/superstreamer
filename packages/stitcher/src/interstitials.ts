import { createUrl } from "./lib/url";
import { getAssetsFromVast } from "./vast";
import type { DateRange } from "./parser";
import type { Session } from "./session";
import type { Interstitial, InterstitialAsset } from "./types";
import type { DateTime } from "luxon";

export function getStaticDateRanges(session: Session, isLive: boolean) {
  return session.interstitials.map<DateRange>((interstitial) => {
    const startDate = interstitial.dateTime;
    const assetListUrl = getAssetListUrl(interstitial, session);

    const clientAttributes: Record<string, number | string> = {
      RESTRICT: "SKIP,JUMP",
      "ASSET-LIST": assetListUrl,
      "CONTENT-MAY-VARY": "YES",
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": getTimelineStyle(interstitial),
    };

    if (!isLive) {
      clientAttributes["RESUME-OFFSET"] = 0;
    }

    const cue: string[] = ["ONCE"];
    if (startDate.equals(session.startTime)) {
      cue.push("PRE");
    }

    if (cue.length) {
      clientAttributes["CUE"] = cue.join(",");
    }

    return {
      classId: "com.apple.hls.interstitial",
      id: `${startDate.toMillis()}`,
      startDate,
      clientAttributes,
    };
  });
}

export async function getAssets(session: Session, dateTime: DateTime) {
  const assets: InterstitialAsset[] = [];

  const interstitial = session.interstitials.find((interstitial) =>
    interstitial.dateTime.equals(dateTime),
  );

  if (interstitial?.vast) {
    const nextAssets = await getAssetsFromVast(interstitial.vast);
    assets.push(...nextAssets);
  }

  if (interstitial?.assets) {
    assets.push(...interstitial.assets);
  }

  return assets;
}

export function appendInterstitials(
  source: Interstitial[],
  interstitials: Interstitial[],
) {
  for (const interstitial of interstitials) {
    const target = source.find((item) =>
      item.dateTime.equals(interstitial.dateTime),
    );

    if (!target) {
      source.push(interstitial);
      continue;
    }

    if (interstitial.assets) {
      if (!target.assets) {
        target.assets = interstitial.assets;
      } else {
        target.assets.push(...interstitial.assets);
      }
    }

    if (interstitial.vast) {
      target.vast = interstitial.vast;
    }

    if (interstitial.assetList) {
      target.assetList = interstitial.assetList;
    }
  }
}

function getAssetListUrl(interstitial: Interstitial, session?: Session) {
  if (interstitial.assetList) {
    return interstitial.assetList.url;
  }
  return createUrl("out/asset-list.json", {
    dt: interstitial.dateTime.toISO(),
    sid: session?.id,
  });
}

function getTimelineStyle(interstitial: Interstitial) {
  if (interstitial.assets) {
    for (const asset of interstitial.assets) {
      if (asset.kind === "ad") {
        return "HIGHLIGHT";
      }
    }
  }

  if (interstitial.vast) {
    return "HIGHLIGHT";
  }

  return "PRIMARY";
}
