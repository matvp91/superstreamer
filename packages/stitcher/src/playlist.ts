import { assert } from "shared/assert";
import { filterMasterPlaylist, formatFilterToQueryParam } from "./filters";
import { getAssets, getStaticDateRanges } from "./interstitials";
import { encrypt } from "./lib/crypto";
import { joinUrl, makeUrl, resolveUri } from "./lib/url";
import {
  parseMasterPlaylist,
  parseMediaPlaylist,
  stringifyMasterPlaylist,
  stringifyMediaPlaylist,
} from "./parser";
import { getRenditions } from "./parser/helpers";
import type { Filter } from "./filters";
import type { Session } from "./session";

export async function formatMasterPlaylist(params: {
  origUrl: string;
  sessionId?: string;
  filter?: Filter;
}) {
  const master = await fetchMasterPlaylist(params.origUrl);

  if (params.filter) {
    filterMasterPlaylist(master, params.filter);
  }

  for (const variant of master.variants) {
    const url = joinUrl(params.origUrl, variant.uri);
    variant.uri = makeMediaUrl({
      url,
      sessionId: params.sessionId,
    });
  }

  const renditions = getRenditions(master.variants);

  renditions.forEach((rendition) => {
    const url = joinUrl(params.origUrl, rendition.uri);
    rendition.uri = makeMediaUrl({
      url,
      sessionId: params.sessionId,
      type: rendition.type,
    });
  });

  return stringifyMasterPlaylist(master);
}

export async function formatMediaPlaylist(
  mediaUrl: string,
  session?: Session,
  renditionType?: string,
) {
  const media = await fetchMediaPlaylist(mediaUrl);

  // We're in a video playlist when we have no renditionType passed along,
  // this means it does not belong to EXT-X-MEDIA, or when we explicitly VIDEO.
  const videoPlaylist = !renditionType || renditionType === "VIDEO";
  const firstSegment = media.segments[0];

  if (session) {
    // If we have a session, we must have a startTime thus meaning we started.
    assert(session.startTime);

    if (media.endlist) {
      assert(firstSegment);
      firstSegment.programDateTime = session.startTime;
    }

    if (videoPlaylist && firstSegment?.programDateTime) {
      // If we have an endlist and a PDT, we can add static date ranges based on this.
      media.dateRanges = getStaticDateRanges(
        firstSegment.programDateTime,
        session,
      );
    }
  }

  media.segments.forEach((segment) => {
    if (segment.map?.uri === "init.mp4") {
      segment.map.uri = joinUrl(mediaUrl, segment.map.uri);
    }
    segment.uri = joinUrl(mediaUrl, segment.uri);
  });

  return stringifyMediaPlaylist(media);
}

export async function formatAssetList(session: Session, timeOffset?: number) {
  const assets = await getAssets(session, timeOffset);
  return {
    ASSETS: assets,
  };
}

async function fetchMasterPlaylist(url: string) {
  const response = await fetch(url);
  const result = await response.text();
  return parseMasterPlaylist(result);
}

async function fetchMediaPlaylist(url: string) {
  const response = await fetch(url);
  const result = await response.text();
  return parseMediaPlaylist(result);
}

export async function fetchDuration(uri: string) {
  const url = resolveUri(uri);
  const variant = (await fetchMasterPlaylist(url))?.variants[0];

  if (!variant) {
    throw new Error(`Missing variant for "${url}"`);
  }

  const media = await fetchMediaPlaylist(joinUrl(url, variant.uri));
  return media.segments.reduce((acc, segment) => {
    acc += segment.duration;
    return acc;
  }, 0);
}

export function makeMasterUrl(params: {
  url: string;
  filter?: Filter;
  session?: Session;
}) {
  const fil = formatFilterToQueryParam(params.filter);

  const outUrl = makeUrl("out/master.m3u8", {
    eurl: encrypt(params.url),
    sid: params.session?.id,
    fil,
  });

  const url = params.session
    ? makeUrl(`session/${params.session.id}/master.m3u8`, {
        fil,
      })
    : undefined;

  return { url, outUrl };
}

function makeMediaUrl(params: {
  url: string;
  sessionId?: string;
  type?: string;
}) {
  return makeUrl("out/playlist.m3u8", {
    eurl: encrypt(params.url),
    sid: params.sessionId,
    type: params.type,
  });
}
