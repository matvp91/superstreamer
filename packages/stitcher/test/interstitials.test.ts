import { describe, expect, test } from "bun:test";
import { mockSessionWithInterstitials } from "./mock";
import { getStaticDateRanges } from "../src/interstitials";

describe("getStaticDateRanges", () => {
  test("should create dateRanges for vod", () => {
    const session = mockSessionWithInterstitials();

    const isLive = false;
    const dateRanges = getStaticDateRanges(session, isLive);

    expect(dateRanges).toMatchSnapshot();
  });

  test("should create dateRanges for live", () => {
    const session = mockSessionWithInterstitials();

    const isLive = true;
    const dateRanges = getStaticDateRanges(session, isLive);

    expect(dateRanges).toMatchSnapshot();
  });
});
