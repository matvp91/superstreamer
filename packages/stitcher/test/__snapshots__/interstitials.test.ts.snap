// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getStaticDateRanges should create dateRanges for vod 1`] = `
[
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A05.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE,PRE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "SPRS-TYPES": "ad",
    },
    "id": "1619950325",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 5,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950325250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A15.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "SPRS-TYPES": "ad",
    },
    "id": "1619950335",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 15,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950335250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A35.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "SPRS-TYPES": "bumper",
    },
    "id": "1619950355",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 35,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950355250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A45.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "SPRS-TYPES": "ad",
    },
    "id": "1619950365",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 45,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950365250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A13%3A45.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
    },
    "id": "1619950425",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 13,
        "month": 5,
        "second": 45,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950425250,
      "weekData": null,
    },
  },
]
`;

exports[`getStaticDateRanges should create dateRanges for live 1`] = `
[
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A05.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE,PRE",
      "RESTRICT": "SKIP,JUMP",
      "SPRS-TYPES": "ad",
    },
    "id": "1619950325",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 5,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950325250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A15.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "SPRS-TYPES": "ad",
    },
    "id": "1619950335",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 15,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950335250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A35.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "SPRS-TYPES": "bumper",
    },
    "id": "1619950355",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 35,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950355250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A45.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "SPRS-TYPES": "ad",
    },
    "id": "1619950365",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 45,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950365250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A13%3A45.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
    },
    "id": "1619950425",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 13,
        "month": 5,
        "second": 45,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950425250,
      "weekData": null,
    },
  },
]
`;
